File: DSA/Manav-Rajput/Day57_question2.java
class Solution {
   
    boolean isPalindrome(Node head) {
        if (head == null || head.next == null) {
            return true; 
        }
        Node slow = head;
        Node fast = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        Node secondHalf = reverse(slow);
        Node firstHalf = head;
        Node firstPointer = firstHalf;
        Node secondPointer = secondHalf;
        boolean isPalindrome = true;
        while (secondPointer != null) {
            if (firstPointer.data != secondPointer.data) {
                isPalindrome = false;
                break;
            }
            firstPointer = firstPointer.next;
            secondPointer = secondPointer.next;
        }

        return isPalindrome;
    }

    Node reverse(Node head) {
        Node prev = null;
        Node current = head;
        while (current != null) {
            Node next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }
        return prev;
    }
}

File: DSA/Arshdeep-Kaur/Day 57/Day57_Question2.cpp
class Solution {
  public:
    Node* getMid(Node* head) {
        Node* slow = head;
        Node* fast = head -> next; 
    
        while (fast != NULL && fast -> next != NULL) {
            fast = fast -> next -> next;
            slow = slow -> next;
        }
    
        return slow;
    }
    
    Node* reverseList(Node* head) {
        Node* prev = NULL;
        Node* curr = head;
        Node* next = NULL;
    
        while (curr != NULL) {
            next = curr -> next;
            curr -> next = prev;
            prev = curr;
            curr = next;
        }
    
        return prev;
    }

    bool isPalindrome(Node *head) {
        if (head -> next == NULL) {
            return true;
        }
        
        Node* mid = getMid(head);
        Node* temp = mid -> next;
        mid -> next = reverseList(temp);
    
        Node* head1 = head;
        Node* head2 = mid -> next;
    
        while (head2 != NULL) {
            if ( (head1 -> data) != (head2 -> data) ) {
                return false;
            }
    
            head1 = head1 -> next;
            head2 = head2 -> next;
        }
    
        temp = mid -> next;
        mid -> next = reverseList(temp);
        return true;
    }
};
